server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -0,0 +1,330 @@
+            "reconcile-desired-balance",
+            new ReconcileDesiredBalanceTask(desiredBalance),
+            ClusterStateTaskConfig.build(Priority.URGENT),
+            executor
+        );
+    }
+
+    protected void reconcile(DesiredBalance desiredBalance, RoutingAllocation allocation) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("Reconciling desired balance: {}", desiredBalance);
+        } else {
+            logger.debug("Reconciling desired balance for [{}]", desiredBalance.lastConvergedIndex());
+        }
+        allocationOrdering.retainNodes(getNodeIds(allocation.routingNodes()));
+        recordTime(cumulativeReconciliationTime, new DesiredBalanceReconciler(desiredBalance, allocation, allocationOrdering)::run);
+    }
+
+    public DesiredBalance getDesiredBalance() {
+        return currentDesiredBalance;
+    }
+
+    public DesiredBalanceStats getStats() {
+        return new DesiredBalanceStats(
+            currentDesiredBalance.lastConvergedIndex(),
+            desiredBalanceComputation.isActive(),
+            computationsSubmitted.count(),
+            computationsExecuted.count(),
+            computationsConverged.count(),
+            desiredBalanceComputer.iterations.sum(),
+            cumulativeComputationTime.count(),
+            cumulativeReconciliationTime.count()
+        );
+    }
+
+    private void onNoLongerMaster() {
+        if (indexGenerator.getAndSet(-1) != -1) {
+            currentDesiredBalance = DesiredBalance.INITIAL;
+            queue.completeAllAsNotMaster();
+            pendingDesiredBalanceMoves.clear();
+            allocationOrdering.clear();
+        }
+    }
+
+    private final class ReconcileDesiredBalanceTask implements ClusterStateTaskListener {
+        private final DesiredBalance desiredBalance;
+
+        private ReconcileDesiredBalanceTask(DesiredBalance desiredBalance) {
+            this.desiredBalance = desiredBalance;
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            assert MasterService.isPublishFailureException(e) : e;
+            onNoLongerMaster();
+        }
+    }
+
+    private final class ReconcileDesiredBalanceExecutor implements ClusterStateTaskExecutor<ReconcileDesiredBalanceTask> {
+
+        @Override
+        public ClusterState execute(BatchExecutionContext<ReconcileDesiredBalanceTask> batchExecutionContext) {
+            var latest = findLatest(batchExecutionContext.taskContexts());
+            var newState = applyBalance(batchExecutionContext, latest);
+            discardSupersededTasks(batchExecutionContext.taskContexts(), latest);
+            return newState;
+        }
+
+        private TaskContext<ReconcileDesiredBalanceTask> findLatest(List<TaskContext<ReconcileDesiredBalanceTask>> taskContexts) {
+            return taskContexts.stream().max(Comparator.comparing(context -> context.getTask().desiredBalance.lastConvergedIndex())).get();
```

Comment:
:+1: tracked in https://github.com/elastic/elasticsearch/issues/91386; @idegtiarenko could you look into this?

Reply:
Not sure I fully understand the question, but I would try to bring the sample to make it easier to discuss:

* There is some initial state with `lastConvergedIndex=1234`
* We receive a create index request. Computation is started but never finished as there is a new create request coming in. We publish a new (not converged) desired balance with `lastConvergedIndex=1234` (index is the same as above, however the balance might be different)
* We process a next create index request. Computation is started and completed. We publish a new (converged) desired balance with `lastConvergedIndex=1235`.

According to above:
* we would like to pick the latest balance as it contains the most up to date state.
* in case there are multiple entries with the highest value (as if only 1 & 2 tasks published), I guess it is okay to pick a random one of them as a newer computation would come up eventually.

Reply:
> in case there are multiple entries with the highest value (as if only 1 & 2 tasks published), I guess it is okay to pick a random one of them as a newer computation would come up eventually.

On closer inspection, I think the argument is different. The `lastConvergedIndex` of all tasks is expected to be unique, since we only ever submit these tasks from the `ContinuousComputation` and only when it is still fresh, i.e., we expect `lastConvergedIndex` to be fresh too. That makes sense to me. Perhaps we can assert that the `lastConvergedIndex` is unique here?